src/NHibernate/Hql/Ast/ANTLR/Exec/AbstractStatementExecutor.cs

```diff
@@ -188,9 +188,9 @@ protected virtual void DropTemporaryTableIfNecessary(IQueryable persister, ISess
 			{
 				IIsolatedWork work = new TmpIdTableDropIsolatedWork(persister, log, session);
 
-				if (ShouldIsolateTemporaryTableDDL())
+				if (ShouldIsolateTemporaryTableDDL() && session.ConnectionManager.CurrentTransaction != null)
 				{
-					session.ConnectionManager.Transaction.RegisterSynchronization(
+					session.ConnectionManager.CurrentTransaction.RegisterSynchronization(
```

Comment:
This is a "side" bug-fix indeed. The temporary table would never have been dropped previously, in case the operation occurred outside of a transaction. The NHibernate transaction would have received the drop to do "after transaction", but the synchs are executed only on actual commit, not upon `ConnectionManager.AfterTransaction` call.

Reply:
This use case got me thinking if it's useful option for someone. If I get it right in old implementation you can supply some after transaction events before actual DB transaction is created.

Reply:
This sounds as quite a corner case.
You mainly could provide "after transaction" events that would have been happily dropped without ever being executed on next non-transacted statement run against the session...
Anyway, the property being just obsoleted, and kept with all its (anti) features, there are no breaking changes here.
Those adapting to use the new property we will have to adapt for starting the transaction first. I do not think there is any actual need for setting after transaction events before starting the transaction, simply because if you want to do anything out of the transaction on the session after setting them, you will lose them.